### ISSUE TITLE
Stringification of multi-char \N{...} in regexps

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14328]
Migrated from [rt.perl.org#123417](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123417) (status was 'resolved')

Searchable as RT123417$

**Comment by p5pRT** [comment]
### From @cpansprout

$ perl5\.12 \-It/lib \-le 'use Cname; print qr/\\N\{foo\}/'
\(?\-xism&#8203;:\\N\{U\+66\.6F\.6F\}\)
$ perl5\.12 \-le 'print "foo" =~ q|\\N\{U\+66\.6F\.6F\}|'
1
$ perl5\.12 \-le 'print "foo" =~ /\\N\{U\+66\.6F\.6F\}/'
Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
Execution of \-e aborted due to compilation errors\.

\(Cname\.pm is in t/lib in the perl source directory\.  It makes \\N\{foo\} resolve to foo\.\)

I get the same results in bleadperl \(pre\-5\.21\.7\)\.

If a regexp stringifies with \\N\{U\+66\.6F\.6F\} and even accepts \\N\{U\+66\.6F\.6F\} syntax as part of a string used as a regexp\, then why can’t I also write /\\N\{U\+66\.6F\.6F\}/?  Should that work?

\(Yes\, you guessed\, this has to do with deparsing\.\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @demerphq

On 12 December 2014 at 07&#8203;:22\, Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org
> wrote&#8203;:

> \# New Ticket Created by  Father Chrysostomos
> \# Please include the string&#8203;:  \[perl \#123417\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123417 >
>
>
> $ perl5\.12 \-It/lib \-le 'use Cname; print qr/\\N\{foo\}/'
> \(?\-xism&#8203;:\\N\{U\+66\.6F\.6F\}\)
> $ perl5\.12 \-le 'print "foo" =~ q|\\N\{U\+66\.6F\.6F\}|'
> 1
> $ perl5\.12 \-le 'print "foo" =~ /\\N\{U\+66\.6F\.6F\}/'
> Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
> Execution of \-e aborted due to compilation errors\.
>
> \(Cname\.pm is in t/lib in the perl source directory\.  It makes \\N\{foo\}
> resolve to foo\.\)
>
> I get the same results in bleadperl \(pre\-5\.21\.7\)\.
>
> If a regexp stringifies with \\N\{U\+66\.6F\.6F\} and even accepts
> \\N\{U\+66\.6F\.6F\} syntax as part of a string used as a regexp\, then why can’t
> I also write /\\N\{U\+66\.6F\.6F\}/?  Should that work?
>
>
IMO yes it should\. The whole point of this transformation is to make it
possible to define a regexp using named escapes in one pattern and scope\,
and have it work in a different scope where that name is not defined\. So
yes this should be fine\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

On 12 December 2014 at 09&#8203;:55\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:

> On 12 December 2014 at 07&#8203;:22\, Father Chrysostomos \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
>> \# New Ticket Created by  Father Chrysostomos
>> \# Please include the string&#8203;:  \[perl \#123417\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123417 >
>>
>>
>> $ perl5\.12 \-It/lib \-le 'use Cname; print qr/\\N\{foo\}/'
>> \(?\-xism&#8203;:\\N\{U\+66\.6F\.6F\}\)
>> $ perl5\.12 \-le 'print "foo" =~ q|\\N\{U\+66\.6F\.6F\}|'
>> 1
>> $ perl5\.12 \-le 'print "foo" =~ /\\N\{U\+66\.6F\.6F\}/'
>> Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
>> Execution of \-e aborted due to compilation errors\.
>>
>> \(Cname\.pm is in t/lib in the perl source directory\.  It makes \\N\{foo\}
>> resolve to foo\.\)
>>
>> I get the same results in bleadperl \(pre\-5\.21\.7\)\.
>>
>> If a regexp stringifies with \\N\{U\+66\.6F\.6F\} and even accepts
>> \\N\{U\+66\.6F\.6F\} syntax as part of a string used as a regexp\, then why can’t
>> I also write /\\N\{U\+66\.6F\.6F\}/?  Should that work?
>>
>>
> IMO yes it should\. The whole point of this transformation is to make it
> possible to define a regexp using named escapes in one pattern and scope\,
> and have it work in a different scope where that name is not defined\. So
> yes this should be fine\.
>

This bug is from toke\.c line 3291 approximately\, which apparently has not
been taught to recognize the dotted char notation\.

&nbsp;                    if \(len == 0 || len \!= \(STRLEN\)\(e \- s\)\) \{
&nbsp;                        yyerror\("Invalid hexadecimal number in \\\\N\{U\+\.\.\.\}"\);
&nbsp;                        s = e \+ 1;
&nbsp;                        continue;
&nbsp;                    \}

I consider this a bug for sure\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Dec 12 02&#8203;:18&#8203;:43 2014\, demerphq wrote&#8203;:
> On 12 December 2014 at 09&#8203;:55\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> 
> > On 12 December 2014 at 07&#8203;:22\, Father Chrysostomos \<
> > perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> >
> >> \# New Ticket Created by  Father Chrysostomos
> >> \# Please include the string&#8203;:  \[perl \#123417\]
> >> \# in the subject line of all future correspondence about this issue\.
> >> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123417 >
> >>
> >>
> >> $ perl5\.12 \-It/lib \-le 'use Cname; print qr/\\N\{foo\}/'
> >> \(?\-xism&#8203;:\\N\{U\+66\.6F\.6F\}\)
> >> $ perl5\.12 \-le 'print "foo" =~ q|\\N\{U\+66\.6F\.6F\}|'
> >> 1
> >> $ perl5\.12 \-le 'print "foo" =~ /\\N\{U\+66\.6F\.6F\}/'
> >> Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
> >> Execution of \-e aborted due to compilation errors\.
> >>
> >> \(Cname\.pm is in t/lib in the perl source directory\.  It makes
> >> \\N\{foo\}
> >> resolve to foo\.\)
> >>
> >> I get the same results in bleadperl \(pre\-5\.21\.7\)\.
> >>
> >> If a regexp stringifies with \\N\{U\+66\.6F\.6F\} and even accepts
> >> \\N\{U\+66\.6F\.6F\} syntax as part of a string used as a regexp\, then why
> >> can’t
> >> I also write /\\N\{U\+66\.6F\.6F\}/?  Should that work?
> >>
> >>
> > IMO yes it should\. The whole point of this transformation is to make
> > it
> > possible to define a regexp using named escapes in one pattern and
> > scope\,
> > and have it work in a different scope where that name is not defined\.
> > So
> > yes this should be fine\.
> >
> 
> This bug is from toke\.c line 3291 approximately\, which apparently has
> not
> been taught to recognize the dotted char notation\.
> 
> if \(len == 0 || len \!= \(STRLEN\)\(e \- s\)\) \{
>     yyerror\("Invalid hexadecimal number in \\\\N\{U\+\.\.\.\}"\);
>     s = e \+ 1;
>     continue;
> \}
> 
> I consider this a bug for sure\.

I’m working on a fix\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @khwilliamson

On 12/12/2014 09&#8203;:29 AM\, Father Chrysostomos via RT wrote&#8203;:
> On Fri Dec 12 02&#8203;:18&#8203;:43 2014\, demerphq wrote&#8203;:
>> On 12 December 2014 at 09&#8203;:55\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
>>
>>> On 12 December 2014 at 07&#8203;:22\, Father Chrysostomos \<
>>> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>>
>>>> \# New Ticket Created by  Father Chrysostomos
>>>> \# Please include the string&#8203;:  \[perl \#123417\]
>>>> \# in the subject line of all future correspondence about this issue\.
>>>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123417 >
>>>>
>>>>
>>>> $ perl5\.12 \-It/lib \-le 'use Cname; print qr/\\N\{foo\}/'
>>>> \(?\-xism&#8203;:\\N\{U\+66\.6F\.6F\}\)
>>>> $ perl5\.12 \-le 'print "foo" =~ q|\\N\{U\+66\.6F\.6F\}|'
>>>> 1
>>>> $ perl5\.12 \-le 'print "foo" =~ /\\N\{U\+66\.6F\.6F\}/'
>>>> Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
>>>> Execution of \-e aborted due to compilation errors\.
>>>>
>>>> \(Cname\.pm is in t/lib in the perl source directory\.  It makes
>>>> \\N\{foo\}
>>>> resolve to foo\.\)
>>>>
>>>> I get the same results in bleadperl \(pre\-5\.21\.7\)\.
>>>>
>>>> If a regexp stringifies with \\N\{U\+66\.6F\.6F\} and even accepts
>>>> \\N\{U\+66\.6F\.6F\} syntax as part of a string used as a regexp\, then why
>>>> can’t
>>>> I also write /\\N\{U\+66\.6F\.6F\}/?  Should that work?
>>>>
>>>>
>>> IMO yes it should\. The whole point of this transformation is to make
>>> it
>>> possible to define a regexp using named escapes in one pattern and
>>> scope\,
>>> and have it work in a different scope where that name is not defined\.
>>> So
>>> yes this should be fine\.
>>>
>>
>> This bug is from toke\.c line 3291 approximately\, which apparently has
>> not
>> been taught to recognize the dotted char notation\.
>>
>> if \(len == 0 || len \!= \(STRLEN\)\(e \- s\)\) \{
>>      yyerror\("Invalid hexadecimal number in \\\\N\{U\+\.\.\.\}"\);
>>      s = e \+ 1;
>>      continue;
>> \}
>>
>> I consider this a bug for sure\.
>
> I’m working on a fix\.
>

I do not consider this a bug\, and I am against changing it\.  This 
notation is a kludge to solve a bug efficiently\.  I deliberately made it 
for internal use only\.

Before exposing it\, there needs to be discussion and agreement on this 
list as to why it should be exposed\.  This means that there needs to be 
a clear and compelling use case for someone needing it\.

I don't think there are any\.  This notation is used for Unicode named 
sequences\.  The whole point of a named sequence is to hide the fact that 
something that appears to be a single character is actually implemented 
as a sequence of multiple ones\.  Behind every official named sequence is 
someone who thought that it should be represented by a single character\, 
but were overruled by various standards committees\, often with much 
gnashing of teeth\.

I don't see why this notation is necessary for deparsing\.  If the 
original name is not retrievable\, the sequence can be expressed as a 
series of characters using \\x\{\} or \\N\{U\+\}\.

Since this notation was internal only\, there was no thought given as to 
whether it is a good one or not\.

And we shouldn't be inventing new syntax unless there is a clear 
benefit\.  I don't see any here\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Dec 12 10&#8203;:13&#8203;:01 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> I do not consider this a bug\, and I am against changing it\.  This 
> notation is a kludge to solve a bug efficiently\.  I deliberately made it 
> for internal use only\.
> 
> Before exposing it\, there needs to be discussion and agreement on this 
> list as to why it should be exposed\.  This means that there needs to be 
> a clear and compelling use case for someone needing it\.
> 
> I don't think there are any\.  This notation is used for Unicode named 
> sequences\.  The whole point of a named sequence is to hide the fact that 
> something that appears to be a single character is actually implemented 
> as a sequence of multiple ones\.  Behind every official named sequence is 
> someone who thought that it should be represented by a single character\, 
> but were overruled by various standards committees\, often with much 
> gnashing of teeth\.
> 
> I don't see why this notation is necessary for deparsing\.  If the 
> original name is not retrievable\, the sequence can be expressed as a 
> series of characters using \\x\{\} or \\N\{U\+\}\.

How do you do that within a character class?

On a related note&#8203;:  It appears that "\\N\{U\+41\}" will treat that as a Unicode codepoint\, even on EBCDIC\, which it will do by recoding it into the native character set\, such that "\\N\{U\+41\}" eq "A" is always true\.  But it appears to me\, from reading the code\, that "" =~ '\\N\{U\+41\}' will match against EBCDIC character 65\, not ASCII/Unicode character 65\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @khwilliamson

On 12/12/2014 01&#8203;:40 PM\, Father Chrysostomos via RT wrote&#8203;:
> On Fri Dec 12 10&#8203;:13&#8203;:01 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
>> I do not consider this a bug\, and I am against changing it\.  This
>> notation is a kludge to solve a bug efficiently\.  I deliberately made it
>> for internal use only\.
>>
>> Before exposing it\, there needs to be discussion and agreement on this
>> list as to why it should be exposed\.  This means that there needs to be
>> a clear and compelling use case for someone needing it\.
>>
>> I don't think there are any\.  This notation is used for Unicode named
>> sequences\.  The whole point of a named sequence is to hide the fact that
>> something that appears to be a single character is actually implemented
>> as a sequence of multiple ones\.  Behind every official named sequence is
>> someone who thought that it should be represented by a single character\,
>> but were overruled by various standards committees\, often with much
>> gnashing of teeth\.
>>
>> I don't see why this notation is necessary for deparsing\.  If the
>> original name is not retrievable\, the sequence can be expressed as a
>> series of characters using \\x\{\} or \\N\{U\+\}\.
>
> How do you do that within a character class?

It doesn't\.  What happens in a character class is that it gets rewritten 
to be \(?&#8203;:\\N\{sequence\}|\[xyz\]\)\.  It would help if you gave me an example 
where you want this\.
>
> On a related note&#8203;:  It appears that "\\N\{U\+41\}" will treat that as a Unicode codepoint\, even on EBCDIC\, which it will do by recoding it into the native character set\, such that "\\N\{U\+41\}" eq "A" is always true\.  But it appears to me\, from reading the code\, that "" =~ '\\N\{U\+41\}' will match against EBCDIC character 65\, not ASCII/Unicode character 65\.
>

Thanks for finding this\.  I'll look into it\.  Whereabouts were you 
looking in the code?

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Dec 12 16&#8203;:29&#8203;:06 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 12/12/2014 01&#8203;:40 PM\, Father Chrysostomos via RT wrote&#8203;:
> > On Fri Dec 12 10&#8203;:13&#8203;:01 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> >> I don't see why this notation is necessary for deparsing\.  If the
> >> original name is not retrievable\, the sequence can be expressed as a
> >> series of characters using \\x\{\} or \\N\{U\+\}\.
> >
> > How do you do that within a character class?
> 
> It doesn't\.  What happens in a character class is that it gets
> rewritten
> to be \(?&#8203;:\\N\{sequence\}|\[xyz\]\)\.  It would help if you gave me an example
> where you want this\.

How do I deparse this program?

$ \./perl \-Ilib \-It/lib \-MCname \-MO=Deparse \-e '/\[\\N\{foo\}\]/'
BEGIN \{
&nbsp;    $^H\{'charnames'\} = 'CODE\(0x7feaf1035398\)';
\}
/\[\\N\{U\+66\.6F\.6F\}\]/;
\-e syntax OK

\(The $^H\{'charnames'\} thing is bogus\, but that is a separate issue\.\)

It looks as though I will have to parse the regexp in B&#8203;::Deparse and rewrite any char classes that contain \\N\{\.\.\.\}\.

> > On a related note&#8203;:  It appears that "\\N\{U\+41\}" will treat that as a
> > Unicode codepoint\, even on EBCDIC\, which it will do by recoding it
> > into the native character set\, such that "\\N\{U\+41\}" eq "A" is always
> > true\.  But it appears to me\, from reading the code\, that "" =~
> > '\\N\{U\+41\}' will match against EBCDIC character 65\, not ASCII/Unicode
> > character 65\.
> >
> 
> Thanks for finding this\.  I'll look into it\.  Whereabouts were you
> looking in the code?

Near this comment in toke\.c&#8203;:

&nbsp;                        /\* On EBCDIC platforms\, in \\N\{U\+\.\.\.\}\, the '\.\.\.' is a
&nbsp;                         \* Unicode value\, so convert to native so downstream
&nbsp;                         \* code can continue to assume it's native \*/

What immediately follows converts the code from Unicode to native\.  It is only reached for pattern literals \(m//\, s///\, qr//\)\.  That means that a string containing literal '\\N\{U\+\.\.\.\}' will bypass the conversion when fed to the regexp engine\.  It seems to me that the regexp engine should do the conversion and convert back upon stringification\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Dec 12 18&#8203;:00&#8203;:40 2014\, sprout wrote&#8203;:
> What immediately follows converts the code from Unicode to native\.  It
> is only reached for pattern literals \(m//\, s///\, qr//\)\.  That means
> that a string containing literal '\\N\{U\+\.\.\.\}' will bypass the
> conversion when fed to the regexp engine\.  It seems to me that the
> regexp engine should do the conversion and convert back upon
> stringification\.

Actually\, converting back should not be necessary\, as the original stringified form is preserved\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @demerphq

On 12 December 2014 at 19&#8203;:12\, Karl Williamson \<public@&#8203;khwilliamson\.com>
wrote&#8203;:

> On 12/12/2014 09&#8203;:29 AM\, Father Chrysostomos via RT wrote&#8203;:
>
>> On Fri Dec 12 02&#8203;:18&#8203;:43 2014\, demerphq wrote&#8203;:
>>
>>> On 12 December 2014 at 09&#8203;:55\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
>>>
>>>  On 12 December 2014 at 07&#8203;:22\, Father Chrysostomos \<
>>>> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>>>
>>>>  \# New Ticket Created by  Father Chrysostomos
>>>>> \# Please include the string&#8203;:  \[perl \#123417\]
>>>>> \# in the subject line of all future correspondence about this issue\.
>>>>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123417 >
>>>>>
>>>>>
>>>>> $ perl5\.12 \-It/lib \-le 'use Cname; print qr/\\N\{foo\}/'
>>>>> \(?\-xism&#8203;:\\N\{U\+66\.6F\.6F\}\)
>>>>> $ perl5\.12 \-le 'print "foo" =~ q|\\N\{U\+66\.6F\.6F\}|'
>>>>> 1
>>>>> $ perl5\.12 \-le 'print "foo" =~ /\\N\{U\+66\.6F\.6F\}/'
>>>>> Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
>>>>> Execution of \-e aborted due to compilation errors\.
>>>>>
>>>>> \(Cname\.pm is in t/lib in the perl source directory\.  It makes
>>>>> \\N\{foo\}
>>>>> resolve to foo\.\)
>>>>>
>>>>> I get the same results in bleadperl \(pre\-5\.21\.7\)\.
>>>>>
>>>>> If a regexp stringifies with \\N\{U\+66\.6F\.6F\} and even accepts
>>>>> \\N\{U\+66\.6F\.6F\} syntax as part of a string used as a regexp\, then why
>>>>> can’t
>>>>> I also write /\\N\{U\+66\.6F\.6F\}/?  Should that work?
>>>>>
>>>>>
>>>>>  IMO yes it should\. The whole point of this transformation is to make
>>>> it
>>>> possible to define a regexp using named escapes in one pattern and
>>>> scope\,
>>>> and have it work in a different scope where that name is not defined\.
>>>> So
>>>> yes this should be fine\.
>>>>
>>>>
>>> This bug is from toke\.c line 3291 approximately\, which apparently has
>>> not
>>> been taught to recognize the dotted char notation\.
>>>
>>> if \(len == 0 || len \!= \(STRLEN\)\(e \- s\)\) \{
>>>      yyerror\("Invalid hexadecimal number in \\\\N\{U\+\.\.\.\}"\);
>>>      s = e \+ 1;
>>>      continue;
>>> \}
>>>
>>> I consider this a bug for sure\.
>>>
>>
>> I’m working on a fix\.
>>
>>
> I do not consider this a bug\, and I am against changing it\.  This notation
> is a kludge to solve a bug efficiently\.  I deliberately made it for
> internal use only\.
>
>
No\, it is a work around for the fact that lexically scope named sequences
 which produce multicharacter output do not play nicely with the principles
of regexp composition\.

While I respect your opinion as the one who implemented it\, I was the one
who suggested in the first place so I feel entitled to disagree\.

Before exposing it\, there needs to be discussion and agreement on this list
> as to why it should be exposed\.  This means that there needs to be a clear
> and compelling use case for someone needing it\.
>

The rules of Perl regexp composition require it\.

We must be able to take any regexp\, as a pattern or a string\, and
concatenate them together in a different context and produce the same
program\. We must be able to take a regexp pattern\, turn it into a string\,
store it somewhere\, and load the string\, put it into a qr// and have it
match the same way it did in the first place\. This is a core part of how
Perl treats regexp patterns\. Arbitrary charname mappings that are lexically
scoped directly conflict with this requirement\, so we have to translate
them into an intermediate form so they work in different scopes\. And if you
can concatenate it as a pattern you can also convert the pattern to a
string\, and then at a later point convert it back to a pattern again
\(possible involving concatenation with other components\)\. That includes via
code generation\.

The example that FC came up with is a  example\. If I write a piece of code\,
and then deparse it to execute in a  different context it should do the
same thing\.

Lexicaly scope named sequences break this rule\. Anyone is allowed to define
any named sequence they choose\. However the regexp's produced in those
scopes need to produce something that can be executed in a different
context\. Therefore we translate it to this different syntax\.



>
> I don't think there are any\.  This notation is used for Unicode named
> sequences\.


That may be how you view it\, the way random hacker sees named sequences is
quite different\. Note that people can implement named sequences that
produce \*random\* bytes\!


> The whole point of a named sequence is to hide the fact that something
> that appears to be a single character is actually implemented as a sequence
> of multiple ones\.  Behind every official named sequence is someone who
> thought that it should be represented by a single character\, but were
> overruled by various standards committees\, often with much gnashing of
> teeth\.
>

Well\, ever since charnames was implemented we have documented that you can
define whatever mapping a developer chooses\. And that leaks into regexp
composition\.


>
> I don't see why this notation is necessary for deparsing\.  If the original
> name is not retrievable\, the sequence can be expressed as a series of
> characters using \\x\{\} or \\N\{U\+\}\.
>
>
Because people can lexically define a named sequence\. We need to preserve
the mapping of name to bytes\, so that when that pattern is used in a
 different context\, without the definition in scope it still works\.


> Since this notation was internal only\, there was no thought given as to
> whether it is a good one or not\.
>
> And we shouldn't be inventing new syntax unless there is a clear benefit\.
> I don't see any here\.
>


There is clear benefit\. Regexp need to be composable\, serializable\, and
executed in different contexts without having to drag around the lexical
context they were defined in\. Its really that simple\.

I feel really strongly that FC's example cases should work Just Fine\.

I consider that this dies to be plain and simply a bug that violates the
rules of regexp pattern/string composition\.

perl \-le'my $str=""; \{use charnames \(\); BEGIN\{ $^H\{charnames\}= sub \{ return
"foo" \};\} $str=""\.qr/\\N\{LATIN CAPITAL LETTER A\}/;\} my $pat="\(foo|$str\)";
eval "\\$thing =~ /$pat/; 1" or warn $@&#8203;'
Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \(eval 1\) line 1\, within pattern

That should Just Work\. I admit this is an annoying dark and crusty part of
perl that probably should not have been created in the first place\, but as
long as we allow customer translators and arbitrary multi\-character
mappings we need this to work\.

There is an alternative\, we could convert the named sequence to a different
representation\, such as hex\. But that would be longer\, and would not have
the same unicode semantics\.

cheers\,
Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 13 December 2014 at 03&#8203;:00\, Father Chrysostomos via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> On Fri Dec 12 16&#8203;:29&#8203;:06 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> > On 12/12/2014 01&#8203;:40 PM\, Father Chrysostomos via RT wrote&#8203;:
> > > On Fri Dec 12 10&#8203;:13&#8203;:01 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> > >> I don't see why this notation is necessary for deparsing\.  If the
> > >> original name is not retrievable\, the sequence can be expressed as a
> > >> series of characters using \\x\{\} or \\N\{U\+\}\.
> > >
> > > How do you do that within a character class?
> >
> > It doesn't\.  What happens in a character class is that it gets
> > rewritten
> > to be \(?&#8203;:\\N\{sequence\}|\[xyz\]\)\.  It would help if you gave me an example
> > where you want this\.
>
> How do I deparse this program?
>
> $ \./perl \-Ilib \-It/lib \-MCname \-MO=Deparse \-e '/\[\\N\{foo\}\]/'
> BEGIN \{
>     $^H\{'charnames'\} = 'CODE\(0x7feaf1035398\)';
> \}
> /\[\\N\{U\+66\.6F\.6F\}\]/;
> \-e syntax OK
>
>
I dont think that makes sense\.


> \(The $^H\{'charnames'\} thing is bogus\, but that is a separate issue\.\)
>
>
I agree it is bogus\. charnames should allow one to do&#8203;:

\{

&nbsp;  use charnames blah\_blah => sub \{ "my blah blah" \};

\}

And I thought it did in the past\.\.\.



> It looks as though I will have to parse the regexp in B&#8203;::Deparse and
> rewrite any char classes that contain \\N\{\.\.\.\}\.
>
>
That is definitely not right\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 13 December 2014 at 14&#8203;:09\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:

> There is an alternative\, we could convert the named sequence to a
> different representation\, such as hex\. But that would be longer\, and would
> not have the same unicode semantics\.
>

For those unversed in the minutae\, if someone sets up a charname mapper so
that \\N\{foo\} produces "I am HAL\." and then a sub in that scope returns a
qr// then we need something like this&#8203;:

my $pat= "\(?&#8203;:$other\_stuff|$pat\)";

to be usable to match the same \(but with unicode semantics\) as&#8203;:

my $pat= "\(?&#8203;:$other\_stuff|I am HAL\\\.\)";

Because we need the unicode semantics\, and because it also avoids secondary
translation issues\, such as treating the "\." as a metacharacter\, and
because it means we have no issues with the scoping of the \\N\{\} definitions
we translate these things into a "dotted hex" notation\.

It is a somewhat ugly solution\, but it seems forced on us by the
interaction of lexically scoped arbitrary multicharacter named sequences\,
regexp composition rules\, deparsing\, eval string\, etc\.

We have had bugs where people used named sequences and they broke in
various external context use cases\. I consider the one that FC originally
posted\, and case like the following&#8203;:

$ perl \-le'my $str=""; \{use charnames \(\); BEGIN\{ $^H\{charnames\}= sub \{
return "foo" \};\} $str=""\.qr/\\N\{LATIN CAPITAL LETTER A\}/;\} my
$pat="\(foo|$str\)"; print $pat; eval "\\$thing =~ /$pat/; 1" or warn $@&#8203;'
\(foo|\(?^u&#8203;:\\N\{U\+66\.6F\.6F\}\)\)
Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \(eval 1\) line 1\, within pattern

to similarly be bugs\.

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 12 December 2014 at 17&#8203;:29\, Father Chrysostomos via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
> On Fri Dec 12 02&#8203;:18&#8203;:43 2014\, demerphq wrote&#8203;:
> > On 12 December 2014 at 09&#8203;:55\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> >
> > > On 12 December 2014 at 07&#8203;:22\, Father Chrysostomos \<
> > > perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > >
> > >> \# New Ticket Created by  Father Chrysostomos
> > >> \# Please include the string&#8203;:  \[perl \#123417\]
> > >> \# in the subject line of all future correspondence about this issue\.
> > >> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123417 >
> > >>
> > >>
> > >> $ perl5\.12 \-It/lib \-le 'use Cname; print qr/\\N\{foo\}/'
> > >> \(?\-xism&#8203;:\\N\{U\+66\.6F\.6F\}\)
> > >> $ perl5\.12 \-le 'print "foo" =~ q|\\N\{U\+66\.6F\.6F\}|'
> > >> 1
> > >> $ perl5\.12 \-le 'print "foo" =~ /\\N\{U\+66\.6F\.6F\}/'
> > >> Invalid hexadecimal number in \\N\{U\+\.\.\.\} at \-e line 1\, within pattern
> > >> Execution of \-e aborted due to compilation errors\.
> > >>
> > >> \(Cname\.pm is in t/lib in the perl source directory\.  It makes
> > >> \\N\{foo\}
> > >> resolve to foo\.\)
> > >>
> > >> I get the same results in bleadperl \(pre\-5\.21\.7\)\.
> > >>
> > >> If a regexp stringifies with \\N\{U\+66\.6F\.6F\} and even accepts
> > >> \\N\{U\+66\.6F\.6F\} syntax as part of a string used as a regexp\, then why
> > >> can’t
> > >> I also write /\\N\{U\+66\.6F\.6F\}/?  Should that work?
> > >>
> > >>
> > > IMO yes it should\. The whole point of this transformation is to make
> > > it
> > > possible to define a regexp using named escapes in one pattern and
> > > scope\,
> > > and have it work in a different scope where that name is not defined\.
> > > So
> > > yes this should be fine\.
> > >
> >
> > This bug is from toke\.c line 3291 approximately\, which apparently has
> > not
> > been taught to recognize the dotted char notation\.
> >
> > if \(len == 0 || len \!= \(STRLEN\)\(e \- s\)\) \{
> >     yyerror\("Invalid hexadecimal number in \\\\N\{U\+\.\.\.\}"\);
> >     s = e \+ 1;
> >     continue;
> > \}
> >
> > I consider this a bug for sure\.
>
> I’m working on a fix\.
>

Just curious if you are still on this subject\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Dec 14 01&#8203;:03&#8203;:42 2014\, demerphq wrote&#8203;:
> On 12 December 2014 at 17&#8203;:29\, Father Chrysostomos via RT \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > I’m working on a fix\.
> >
> 
> Just curious if you are still on this subject\.

I threw away my first attempt on seeing Karl’s first response\.  \(Not much loss; it needed a rewrite anyway\.\)  I’ve just been waiting to see if Karl has any response to your responses\.  But I still have it all in my head right now\, so this would be a good time for me to work on it\.  Karl?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @khwilliamson

On 12/14/2014 07&#8203;:23 AM\, Father Chrysostomos via RT wrote&#8203;:
> On Sun Dec 14 01&#8203;:03&#8203;:42 2014\, demerphq wrote&#8203;:
>> On 12 December 2014 at 17&#8203;:29\, Father Chrysostomos via RT \<
>> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>> I’m working on a fix\.
>>>
>>
>> Just curious if you are still on this subject\.
>
> I threw away my first attempt on seeing Karl’s first response\.  \(Not much loss; it needed a rewrite anyway\.\)  I’ve just been waiting to see if Karl has any response to your responses\.  But I still have it all in my head right now\, so this would be a good time for me to work on it\.  Karl?
>

I have only time for a partial response\.

It is not ok to change the syntax of Perl without properly vetting it on 
this list\.  That means it has to be a non\-hasty decision that more than 
the three of us have to weigh in on\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Dec 14 06&#8203;:23&#8203;:44 2014\, sprout wrote&#8203;:
> I threw away my first attempt on seeing Karl’s first response\.  \(Not
> much loss; it needed a rewrite anyway\.\)  I’ve just been waiting to see
> if Karl has any response to your responses\.  But I still have it all
> in my head right now\, so this would be a good time for me to work on
> it\.

I have a patch waiting on the sprout/123417 branch\.  It’s sprout/123417^\, \*not\* the tip of the branch\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @davidnicol

For what it's worth\, it has long seemed to me that having regex
stringification and interpolation as the way that interpolation of
predefined regular expressions into regular expression strings works
is fragile\, and should be replaced by regular expression interpolation
being different from string interpolation\, in much the same way that
eval\-block is different from eval\-string\.

A coderef that closes over a lexical may be used in a different
context and when it is deparsed for human eyes the association of the
variables is lost\. There is no expectation that taking a string\-eval
of a deparsed coderef will provide a perfectly functioning
replacement\. The analogous expectation with regular expressions\, which
now seem to have their own lexical issues aside from those raised with
allowing embedded coderefs\, seems wrong\, and therefore

it might be better to fix this by refactoring regex interpolation to
operate at a deeper level where no deparsing and remapping is
required\, rather than further adjusting regex\-to\-string to leak remote
lexicals\.

thanks for reading

dln

\-\- 
Mmm\, jelly donut\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Dec 15 10&#8203;:38&#8203;:30 2014\, davidnicol@&#8203;gmail\.com wrote&#8203;:
> For what it's worth\, it has long seemed to me that having regex
> stringification and interpolation as the way that interpolation of
> predefined regular expressions into regular expression strings works
> is fragile\, and should be replaced by regular expression interpolation
> being different from string interpolation\, in much the same way that
> eval\-block is different from eval\-string\.
> 
> A coderef that closes over a lexical may be used in a different
> context and when it is deparsed for human eyes the association of the
> variables is lost\. There is no expectation that taking a string\-eval
> of a deparsed coderef will provide a perfectly functioning
> replacement\. The analogous expectation with regular expressions\, which
> now seem to have their own lexical issues aside from those raised with
> allowing embedded coderefs\, seems wrong\, and therefore
> 
> it might be better to fix this by refactoring regex interpolation to
> operate at a deeper level where no deparsing and remapping is
> required\, rather than further adjusting regex\-to\-string to leak remote
> lexicals\.

For regexp code blocks\, we already take existing ops and combine them in /$qr1$qr2/ where $qr1 and 2 have \(?\{\.\.\.\}\) closures in them\.  Whether we should follow that approach to the fullest extent in combining compiled lists of regnodes instead of reparsing them I cannot really say\, as I am no expert on regexp compilation\.  I just think it would be a lot of work\, considering that it took a year for Dave Mitchell to fix code blocks\.

Regardless of that\, stringification and reparsing should work to whatever extent is possible\.  Obviously code block closures can’t work that way\, but everything else should be able to\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Dec 14 09&#8203;:40&#8203;:33 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> It is not ok to change the syntax of Perl without properly vetting it
> on
> this list\.  That means it has to be a non\-hasty decision that more
> than
> the three of us have to weigh in on\.

I think that introducing an internally\-used syntax that the regular expression parser understands falls under the same category as introducing a syntax that allows eval "/$qr/" to work\.  The former needs just as much vetting as the latter\.  The two go together\.  Hence\, I don’t see dotted\-hex support in the lexer as being any more hasty than dotted\-hex support in the regexp parser \(which has already happened\)\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @davidnicol

On Mon\, Dec 15\, 2014 at 7&#8203;:52 PM\, Father Chrysostomos via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
>
> Regardless of that\, stringification and reparsing should work to whatever
> extent is possible\.  Obviously code block closures can’t work that way\, but
> everything else should be able to\.
>

But code block closures \*could\* work that way if we wanted them to\. All
that would be needed is a way for\, say\,

&nbsp;   $\{'SCALAR\(0x18831420\)'\}

to dereference the thing that stringified to that representation\, so both
of these would print 44 on the last line&#8203;:

$ perl \-le 'print $r = \\\($x=43\); print $\{$r\} = 44; print $x'
SCALAR\(0x11030470\)
44
44
$ perl \-le 'print $r = \\\($x=43\); print $\{"$r"\} = 44; print $x'
SCALAR\(0x12c59470\)
44
43
$

It would just take reservation of one global\, for instance\,
@&#8203;core&#8203;::regex&#8203;::precompiled\, that would hold
troublesome precompiled regexes\, then they could get stringified as
$core&#8203;::regex&#8203;::precompiled\[22\] instead
of reproducing their source\, and that would work\, at least until trying to
send the regex across a process membrane\.



\-\- 
Mmm\, jelly donut\.

**Comment by p5pRT** [comment]
### From @ap

\* David Nicol \<davidnicol@&#8203;gmail\.com> \[2014\-12\-16 19&#8203;:05\]&#8203;:
> All that would be needed is a way for\, say\,
>
>    $\{'SCALAR\(0x18831420\)'\}
>
> to dereference the thing that stringified to that representation

Out\, you demon\.

**Comment by p5pRT** [comment]
### From @demerphq

On 16 December 2014 at 02&#8203;:55\, Father Chrysostomos via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
> On Sun Dec 14 09&#8203;:40&#8203;:33 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> > It is not ok to change the syntax of Perl without properly vetting it
> > on
> > this list\.  That means it has to be a non\-hasty decision that more
> > than
> > the three of us have to weigh in on\.
>
> I think that introducing an internally\-used syntax that the regular
> expression parser understands falls under the same category as introducing
> a syntax that allows eval "/$qr/" to work\.  The former needs just as much
> vetting as the latter\.  The two go together\.  Hence\, I don’t see dotted\-hex
> support in the lexer as being any more hasty than dotted\-hex support in the
> regexp parser \(which has already happened\)\.
>

I think this boat has sailed anyway\, the syntax is out there\, and we need
to make code that should work\, work\.

A pattern compiled under 5\.18\.2 that is serialized as a string and then
loaded in 5\.22 needs to work the same way it did originally\.

For me this was never about an "internal syntax"\, when I proposed this
syntax to work around the unfortunate interaction of lexcially scoped
multicharacter arbitrary named escapes and regex composition rules I had
full expectation that the syntax would be valid to a compiled program\, as
it is a requirement of the composition rules of regexes in Perl\.

Yves

**Comment by p5pRT** [comment]
### From @khwilliamson

On 12/17/2014 03&#8203;:40 AM\, demerphq wrote&#8203;:
> A pattern compiled under 5\.18\.2 that is serialized as a string and then
> loaded in 5\.22 needs to work the same way it did originally\.

You're right\.  I reluctantly withdraw my objection\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Dec 17 12&#8203;:32&#8203;:33 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 12/17/2014 03&#8203;:40 AM\, demerphq wrote&#8203;:
> > A pattern compiled under 5\.18\.2 that is serialized as a string and then
> > loaded in 5\.22 needs to work the same way it did originally\.
> 
> You're right\.  I reluctantly withdraw my objection\.

I have applied my patch in 4cbd7e223\.  Concerning the '\\N\{U\+41\}' vs /\\N\{U\+41\}/ EBCDIC issue\, are you still planning to look into it?  Or should I?  \(I’m just trying to avoid duplicate effort\.\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @khwilliamson

On 12/17/2014 06&#8203;:25 PM\, Father Chrysostomos via RT wrote&#8203;:
> On Wed Dec 17 12&#8203;:32&#8203;:33 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
>> On 12/17/2014 03&#8203;:40 AM\, demerphq wrote&#8203;:
>>> A pattern compiled under 5\.18\.2 that is serialized as a string and then
>>> loaded in 5\.22 needs to work the same way it did originally\.
>>
>> You're right\.  I reluctantly withdraw my objection\.
>
> I have applied my patch in 4cbd7e223\.  Concerning the '\\N\{U\+41\}' vs /\\N\{U\+41\}/ EBCDIC issue\, are you still planning to look into it?  Or should I?  \(I’m just trying to avoid duplicate effort\.\)
>

I have it already fixed in a branch

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Dec 17 19&#8203;:31&#8203;:57 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 12/17/2014 06&#8203;:25 PM\, Father Chrysostomos via RT wrote&#8203;:
> > On Wed Dec 17 12&#8203;:32&#8203;:33 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> >> On 12/17/2014 03&#8203;:40 AM\, demerphq wrote&#8203;:
> >>> A pattern compiled under 5\.18\.2 that is serialized as a string and
> >>> then
> >>> loaded in 5\.22 needs to work the same way it did originally\.
> >>
> >> You're right\.  I reluctantly withdraw my objection\.
> >
> > I have applied my patch in 4cbd7e223\.  Concerning the '\\N\{U\+41\}' vs
> > /\\N\{U\+41\}/ EBCDIC issue\, are you still planning to look into it?  Or
> > should I?  \(I’m just trying to avoid duplicate effort\.\)
> >
> 
> I have it already fixed in a branch

This remaining issue was fixed in af352bf2fad22f and b6d67071cc\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @khwilliamson

I'm reopening this ticket because I think this commit was a mistake

================================================================================
commit b7e6151cbc27942266f736d4706550bc35a29639
Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;:   Thu Dec 18 22&#8203;:10&#8203;:49 2014 \-0800

&nbsp;    Also allow /\\N\{\}/

&nbsp;    See 4cbd7e223 and ticket \#123417\.

&nbsp;    If a charnames handler returns the empty string for a particular name\,
&nbsp;    then the regular expression stringifies with an empty \\N\{\} in it\.
&nbsp;    This needs to round\-trip properly \(eval "/$qr/"\)\, just like \\N\{U\+\.\.\.\}\. 
================================================================================

\\N\{\} doesn't make much sense\.  Most likely someone specifying it is not intending it to be what it now is taken as\, a fancy way of saying&#8203;: nothing\.  That means that this silently now allows bugs\.  Prior to this commit\, it was a syntax error in normal input\.  By making it legal\, downstream code in the regex compiler has to now cope with something it never saw before\, and we are still seeing bugs as a result\.  I had added compensatory code there\, not realizing that this was recently changed behavior\, but there were things I missed\.  And Yves knew in the back of his mind that this didn't use to be allowed\, so I went searching\.

The reason for the commit appears to be that we allowed an empty \\N\{\} if the code had installed a custom charnames handler\, and so that needed to be deparsed properly\.  But it would be better to just disallow this nonsensical result from the handler\, than to allow nonsensical stuff everywhere\.

As I write this\, we are after the visible 5\.24 changes freeze\, and close to the full code freeze\.  I however\, don't think this should be allowed as\-is for another release\.  It's been this way only for 5\.22\.  My guess is that the best way forward is to add a deprecation warning in 5\.24 for an empty \\N\{\}\. 

 \-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @rjbs

\* Karl Williamson via RT \<perlbug\-followup@&#8203;perl\.org> \[2016\-03\-16T17&#8203;:36&#8203;:00\]
> As I write this\, we are after the visible 5\.24 changes freeze\, and close to
> the full code freeze\.  I however\, don't think this should be allowed as\-is
> for another release\.  It's been this way only for 5\.22\.  My guess is that the
> best way forward is to add a deprecation warning in 5\.24 for an empty \\N\{\}\. 

This seems reasonable to me\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @khwilliamson

Now deprecated by d8d26cacf9f6b2fbd47665039d13384d3ccfd516
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thank you for submitting this report\.  You have helped make Perl better\.
&nbsp;    
With the release of Perl 5\.24\.0 on May 9\, 2016\, this and 149 other issues have been resolved\.

Perl 5\.24\.0 may be downloaded via https://metacpan.org/release/RJBS/perl-5.24.0

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
